rbx_reflection_database/src/lib.rs

```diff
@@ -1,23 +1,148 @@
+//! but also provides an API for dependents to get a reflection database
+//! from a consistent location.
+//!
+//! The general way this crate should be used is via [`get`]. This method will
+//! search for a locally stored reflection database and return it if it's
+//! found. If it isn't, it will instead return the bundled one.
+//!
+//! Additionally, this crate exposes [`get_local`] and [`get_bundled`] for
+//! only loading the locally stored database or only the bundled one
+//! respectively.
+//!
+//! ## Local Details
+//!
+//! This crate will load a reflection database from the file system if one
+//! exists in the default location. This location varies upon the OS and is
+//! specified here:
+//!
+//! | OS      | Location                                                            |
+//! |:--------|:--------------------------------------------------------------------|
+//! | Windows | `%localappdata%/.rbxreflection/database.msgpack`                    |
+//! | MacOS   | `$HOME/Library/Application Support/.rbxreflection/database.msgpack` |
+//! | Linux   | `$HOME/.rbxreflection/database.msgpack`                             |
+//!
+//! Additionally, a location override may be specified via the `RBX_DATABASE`
+//! environmental variable.
+//!
+//! Both the default `database.msgpack` files and any files pointed to by
+//! `RBX_DATABASE` must be valid MessagePack serializations of a
+//! [`ReflectionDatabase`] if they're present.
 use rbx_reflection::ReflectionDatabase;
 
+use std::{env, fs, path::PathBuf};
+
 static ENCODED_DATABASE: &[u8] = include_bytes!("../database.msgpack");
 
+/// The name of an environmental variable that may be used to specify
+/// the location of a reflection database to use. The expected format of
+/// a file at this point is MessagePack.
+pub const OVERRIDE_PATH_VAR: &str = "RBX_DATABASE";
+
+/// The name of the directory used for the local location for a reflection
+/// database. The directory will be placed inside the current user's
+/// local data folder on MacOS and Windows and inside
+/// the home directory on Linux.
+pub const LOCAL_DIR_NAME: &str = ".rbxreflection";
+
 lazy_static::lazy_static! {
-    static ref DATABASE: ReflectionDatabase<'static> = {
+    static ref BUNDLED_DATABASE: ReflectionDatabase<'static> = {
+        log::debug!("Loading bundled reflection database");
         rmp_serde::decode::from_slice(ENCODED_DATABASE).unwrap_or_else(|e| panic!("could not decode reflection database because: {}", e))
     };
+
+    static ref LOCAL_DATABASE: Option<ReflectionDatabase<'static>> = {
+        let location = get_local_location()?;
+        if let Ok(file) = fs::read(&location) {
+            log::debug!("Loading local reflection database from {}", location.display());
+            Some(
+                rmp_serde::decode::from_slice(&file).unwrap_or_else(|e| {
+                    panic!("could not decode reflection database because: {}", e)
```

Comment:
I did actually consider just using a a `Result` rather than panicking for exactly this reason. It's not an unfounded worry because we've seen problems with a file exactly like this for Selene.

The reason I didn't commit to it out of the gate was because I didn't want to force the `get` to suddenly be a fallible operation, but I suppose it actually is now and lying isn't doing anyone favors. I'm willing to commit to it now so we only have to break things once.